### PR TITLE
Feature/documentation foundation

### PR DESCRIPTION
**Pull request by jaapio**
First part of implementation of the documentation foundation. Contains documention and repository classes.  And a number of interfaces required for the documentgroups.

### CODE REVIEW

File: src/phpDocumentor/Documentation.php [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+    /**
+     * Version number of the documentation.
+     *
+     * @var VersionNumber
+     */
+    private $versionNumber;
+
+    /**
+     * @var DocumentGroup[]
```
Comment (mvriel) [inline src/phpDocumentor/Documentation.php]: Missing a Summary

File: src/phpDocumentor/DocumentationFactory.php [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Factory to create Documentation objects out of Project\Version\Definition.
+ * Will use DocumentationGroupFactories to create the DocumentationGroups.
+ * DocumentationGroupFactories can be registered using the addDocumentationGroupFactory method
+ */
+final class DocumentationFactory
+{
+    /**
+     * @var DocumentGroupFactory[]
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: Do you think there should be a summary here?

File: src/phpDocumentor/DocumentationFactory.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    private $documentGroupFactories;
+
+    /**
+     * Creates Documentation object.
+     *
+     * @param Definition $versionDefinition
+     * @return Documentation
+     * @throws Exception
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: Please mention when the exception is thrown in the description of the tag

File: src/phpDocumentor/DocumentationFactory.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        return new Documentation(null, $versionDefinition->getVersionNumber(), $documentGroups);
+    }
+
+    /**
+     * Find matching factory and use it to create DocumentGroup.
+     *
+     * @param DocumentGroupDefinition $definition
+     * @return DocumentGroup
+     * @throws Exception
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: Please mention when the exception is thrown in the description of the tag

File: src/phpDocumentor/DocumentationFactory.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+     */
+    private function createDocumentGroup(DocumentGroupDefinition $definition)
+    {
+        foreach ($this->documentGroupFactories as $factory) {
+            if ($factory->matches($definition)) {
+                return $factory->create($definition);
+            }
+        }
+
+        throw new Exception('No factory matches document group');
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: Would it be interesting to add a specific Exception class for this?

File: src/phpDocumentor/DocumentationFactory.php [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    /**
+     * Add DocumentGroupFactory to internal register for use during creation of Documentation object.
+     *
+     * @param $documentGroupFactory
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: missing the `@return void`

File: src/phpDocumentor/DocumentationFactory.php [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+     * Find matching factory and use it to create DocumentGroup.
+     *
+     * @param DocumentGroupDefinition $definition
+     * @return DocumentGroup
+     * @throws Exception
+     */
+    private function createDocumentGroup(DocumentGroupDefinition $definition)
+    {
+        foreach ($this->documentGroupFactories as $factory) {
+            if ($factory->matches($definition)) {
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationFactory.php]: Looking good :) :+1:

File: src/phpDocumentor/DocumentationRepository.php [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+ * PHP Version 5.5
+ *
+ * @copyright 2010-2015 Mike van Riel / Naenius (http://www.naenius.com)
+ * @license   http://www.opensource.org/licenses/mit-license.php MIT
+ * @link      http://phpdoc.org
+ */
+
+namespace phpDocumentor;
+use phpDocumentor\Project\VersionNumber;
+use Stash\Pool;
```
Comment (mvriel) [inline src/phpDocumentor/DocumentationRepository.php]: You are absolutely right; I just re-read that section of the docs and the Pool item is more appropriate

File: tests/unit/phpDocumentor/DocumentationFactoryTest.php [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+        $versionDefinition = new Definition(new VersionNumber('1.0.0'));
+
+        $documentation = $this->fixture->create($versionDefinition);
+
+        $this->assertEquals(new VersionNumber('1.0.0'), $documentation->getVersionNumber());
+    }
+
+    /**
+     * @covers ::create
+     * @covers ::<private>
```
Comment (mvriel) [inline tests/unit/phpDocumentor/DocumentationFactoryTest.php]: I did not know about these constructs; thanks! I will apply them as well in my tests

File: tests/unit/phpDocumentor/DocumentationRepositoryTest.php [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+     */
+    private $fixture;
+
+    /**
+     * @var m\Mock
+     */
+    private $cacheMock;
+
+    /**
+     *
```
Comment (mvriel) [inline tests/unit/phpDocumentor/DocumentationRepositoryTest.php]: Empty DocBlock

**Comment by mvriel** [comment]
Absolutely nothing exciting found during the code review; only a few DocBlock comments and one instance where I think that a more specific Exception would be nice.

In other words: excellent work!

**Comment by mvriel** [comment]
@jaapio I just noticed: composer install fails to function on wercker; is there an issue with the dependencies?

**Comment by jaapio** [comment]
All done, ready for merge

**Comment by mvriel** [comment]
:+1: No further comments!

**Pull request merged**